src/components/hmi_message_handler/src/mb_controller.cc

```diff
@@ -0,0 +1,479 @@
+      if(ec) {
+         std::cerr << "ErrorMessage Cancel: " << ": " << ec.message() << "\n";
+      }
+      acceptor_.close(ec);
+      if(ec) {
+         std::cerr << "ErrorMessage Close: " << ": " << ec.message() << "\n";
+      }
+      ioc_.stop();
+   }
+
+   std::string CMessageBrokerController::getMethodName(std::string& method) {
+      std::string return_string="";
+      if(method != "") {
+         int position = method.find(".");
+         if(position != -1) {
+            return_string = method.substr(position + 1);
+         }
+      }
+      return return_string;
+   }
+
+   std::string CMessageBrokerController::GetComponentName(std::string& method) {
+      std::string return_string="";
+      if(method != "") {
+         int position = method.find(".");
+         if(position != -1) {
+            return_string = method.substr(0, position);
+         }
+      }
+      return return_string;
+   }
+
+   bool CMessageBrokerController::addController(WebsocketSession* ws_session, std::string name){
+      bool result = false;
+      std::map <std::string, WebsocketSession*>::iterator it;
+
+      sync_primitives::AutoLock lock(mControllersListLock);
+      it = mControllersList.find(name);
+      if (it == mControllersList.end()) {
+         mControllersList.insert(std::map <std::string, WebsocketSession*>::value_type(name, ws_session));
+         result = true;
+      } else {
+         DBG_MSG(("Controller already exists!\n"));
+      }
+
+      DBG_MSG(("Count of controllers: %zu\n", mControllersList.size()));
+      return result;   
+   }
+
+   void CMessageBrokerController::deleteController(WebsocketSession* ws_session) {
+      {
+         sync_primitives::AutoLock lock(mControllersListLock);
+         std::map <std::string, WebsocketSession*>::iterator it;
+         for(it = mControllersList.begin(); it != mControllersList.end(); it++) {
+            if(it->second == ws_session) {
+               mControllersList.erase(it);
+            }
+         }
+      }
+      removeSubscribersBySession(ws_session);
+   }
+
+   void CMessageBrokerController::deleteController(std::string name) {
+      std::map <std::string, WebsocketSession*>::iterator it;
+      WebsocketSession* ws;
+      {
+         sync_primitives::AutoLock lock(mControllersListLock);
+         it = mControllersList.find(name);
+         if (it != mControllersList.end())
+         {
+            ws = it->second;
+            mControllersList.erase(it);
+         } else {
+            return;
+         }
+      }
+      removeSubscribersBySession(ws);      
+   }
+
+   void CMessageBrokerController::removeSubscribersBySession(const WebsocketSession* ws) {
+      sync_primitives::AutoLock lock(mSubscribersListLock);
+      std::multimap <std::string, WebsocketSession*>::iterator it_s = mSubscribersList.begin();
+      for (; it_s !=mSubscribersList.end(); ) {
+         if (it_s->second == ws) {
+            mSubscribersList.erase(it_s++);
```

Comment:
This iterator is invalidated as soon as erase finishes, I don't believe that incrementing it inside of the statement fixes this.